feat: add report output

wbollock opened the pull request:
This adds a `--report ` flag that is used to summarize tagging changes.
The report consists of a list of Linode IDs affected by tagger, a count
of those Linodes, what tags were added/removed and is sent to stdout.

wbollock committed:
feat: add report output

This adds a `--report ` flag that is used to summarize tagging changes.
It uses a "pretty" table output to state changes in a human readable
way.

Co-authored-by: TJ Hoplock <33664289+tjhop@users.noreply.github.com>

tjhop commented on tagger.go:
```diff
@@ -55,9 +59,15 @@ func newLinodeClient() linodego.Client {
 }
 
-func checkLinodeTagsAgainstConfig(linodes []linodego.Instance, rules []TagRule) (instanceTagMap, error) {
+func checkLinodeTagsAgainstConfig(linodes []linodego.Instance, rules []TagRule) (instanceTagMap, reportMap, error) {
```
Adding a new return value and populating this data here feels a bit forced, IMO. 

Since this function doesn't actually update tags, how would you feel about this being being refactored to leave the `checkLinodeTagsAgainstConfig()` func alone, and instead leveraging a new function to do a diff of the returned `instanceTagMap` vs the instance tags?

wbollock replied:
e09c415f710fcdaa6eb9b5bd35da27b586dab909

tjhop commented on tagger.go:
```diff
@@ -68,22 +78,36 @@ func checkLinodeTagsAgainstConfig(linodes []linodego.Instance, rules []TagRule)
+					} else {
+						// if this tag *is* on the `absent` list, we'll be removing it later
+						absentReportCounter += 1
+						report[tag]["absent"] = strconv.Itoa(absentReportCounter)
```
Is there a reason we're storing this counter as a string in the report map? Can it remain an int in the map? 🤔 
```suggestion
						report[tag]["absent"]++
```

tjhop commented on tagger.go:
```diff
@@ -68,22 +78,36 @@ func checkLinodeTagsAgainstConfig(linodes []linodego.Instance, rules []TagRule)
+						absentReportCounter += 1
+						report[tag]["absent"] = strconv.Itoa(absentReportCounter)
+						// use a list to capture updated LinodeIDs that later get shoved into a map value
+						linodeList = append(linodeList, linode.Label)
```
If we're flattening this list down into a comma separated list after anyway, this could be made a bit more efficient using a string builder:
https://pkg.go.dev/strings#Builder
https://www.calhoun.io/concatenating-and-building-strings-in-go/

wbollock replied:
39e2a3ccc74ca91bd01fe06f02a3c6d23464f2df

tjhop commented on tagger.go:
```diff
@@ -68,22 +78,36 @@ func checkLinodeTagsAgainstConfig(linodes []linodego.Instance, rules []TagRule)
+
+						presentReportCounter += 1
+						linodeList = append(linodeList, linode.Label)
+						report[tag]["present"] = strconv.Itoa(presentReportCounter)
```
Similar comments to the absent tags

tjhop commented on tagger.go:
```diff
@@ -134,6 +157,29 @@ func updateAllInstanceTags(ctx context.Context, client linodego.Client, tagMap i
+			report[tag]["present_linodes"] = report[tag]["present_linodes"][:45] + "..."
+			t.AppendRow(table.Row{tag, report[tag]["present"], report[tag]["present_linodes"], "Added"})
+		}
+	}
```
Since `report` is a map, using `range` with a single iterator like this causes range to iterate over _just_ the keys in the map.
But then we use the map variable at the index of the key during the iteration anyway. Would it be clearer/easier to iterate over both the keys _and_ values? ie, `for tag, data := range report {`

tjhop commented on tagger.go:
```diff
@@ -134,6 +157,29 @@ func updateAllInstanceTags(ctx context.Context, client linodego.Client, tagMap i
+			// truncate linode list lag arbitrarily as it can get too long
+			// column maxwidth and row restrictions don't work well for this
+			report[tag]["absent_linodes"] = report[tag]["absent_linodes"][:45] + "..."
+			t.AppendRow(table.Row{tag, report[tag]["absent"], report[tag]["absent_linodes"], "Removed"})
```
This feels like an unnecessary re-assignment in the map to truncate the results.
Can we change this to have the call to `AppendRow()` pass the slice index range instead of re-assigning it in the map? 

```suggestion
                         // needs `go fmt`'ing
			t.AppendRow(table.Row{
			    tag, 
			    report[tag]["absent"], 
			    report[tag]["absent_linodes"][:45] + "...", 
			    "Removed",
			})
```

tjhop commented on tagger.go:
```diff
@@ -170,6 +216,7 @@ func main() {
 	// prep and parse flags
 	flag.String("config", "", "Path to configuration file to use")
 	flag.String("logging.level", "", "Logging level may be one of: trace, debug, info, warning, error, fatal and panic")
+	flag.Bool("report", true, "Optional report output to summarize tag changes")
```
If we're going to have a flag that can enable report style output, I think we should probably default it to `false`. Otherwise, why make it a flag?

tjhop commented on tagger.go:
```diff
@@ -238,6 +285,12 @@ func main() {
 	}
 
+	// set report status
+	reportFlag := viper.GetString("report")
+	if reportFlag != "" {
```
```suggestion
	if viper.GetBool("report") {
```

The report flag is a boolean, let's use it like one!

tjhop reviewed (commented):
I like the end result! I left a few comments, but I like the direction this is heading 👍

wbollock committed:
feat: add json flag

Outputting our tag changes in a json encoded format allows great
flexibility to parse or consume linode-tagger output in a
non-opinionated manner (as opposed to `--report`).

wbollock committed:
ref!: pipe all logging to stderr

Sending all logs to stderr makes linode-tagger's output much simplier to
pipe to jq from `--json`, or to redirect the `--report` to a file.

I also considered only sending Info/Debug to stderr if report/json flags
were used, but figured it was a worthwhile change regardless. I'm open
to changing this though!

tjhop committed:
ref: remove slices and reflect

* ref: remove the need for pointers in building report

If we allocate structs per iteration, we don't need to do this magic
pointer manipulation.

* fix: sort slices before equality checks

With `slices.Equal`/`reflect.DeepEqual`, if 2 slices have the same
length and same literal elements but in a different order (ie, [1, 2, 3,
4] and [1, 2, 4, 3], they will be considered not equal.

This sorts slices prior to equality checks to ensure that they are
compared accurately.

* ref: reflect pkg no longer needed

The only remaining equality checks we're doing is with slices, for which
we're already using the experimental slices package.

* ref: only return report map in buildReport()

We don't do anything to actually return an error in this function.

* chore: update logging around report generation

tjhop reviewed (approved):
Nice! I think some of the struct usage can be refined a bit, but works as expected 👍
